### ISSUE TITLE
Cannot 2-way sync when using two copies of OGCS on two PCs

### ISSUE THREAD

**Issue by dswhite42** [phw198/OutlookGoogleCalendarSync#382]
### Development Tasklist

**Outlook Calendars :arrow_right: Single Google Calendar**
- [x] Build in ability to store multiple sets of Outlook IDs against a Google item (one for each source calendar)  

**Google Calendars :arrow_right: Single Outlook Calendar**
- [x] Build in ability to store multiple sets of Google IDs against an Outlook item (one for each source calendar)
- [x] Testing

:package: Development build [v2.7.5.7.zip](https://github.com/phw198/OutlookGoogleCalendarSync/files/2300260/v2.7.5.7.zip) available - [deploy as a hotfix](https://github.com/phw198/OutlookGoogleCalendarSync/wiki/Reporting-Problems#applying-a-hotfix) over v2.7.5.x

-------------
**OGCS Version**: v2.6.0.0
**Installed or Portable**: Installed

**Problem Description**  

Hello.  I would like to be able to have OGCS running on both my home PC and work PC, so that if I add entries on my home Outlook, they will go to GoogleCal and then be synced onto my work PC.  (and vice versa).  However, is this an unsupported/impossible setup? 

Here is what I have tried:

1) On the home PC, OGCS is setup for 2-way sync'ing (Outlook <-> Google), using the default mailbox.  Sync options:
    - Merge with existing entries
    - Confirm deletions
    - 0 days in the past  (for testing)
    - 1 day in the future (for testing)
    - 10 minute interval
    - Push Outlook changes immediately

   Test button shows format "g" with 0 entries between 11/4/2017 and 11/6/2017.  So far so good.

2) Create a new Outlook entry on the home PC (11:45pm - 11:48PM, "Test entry #1")

3) Start Sync.  SUCCESS: entry created in GoogleCal.

4) Install OGCS on the work PC.  Again, setup for 2-way sync'ing.  This time use the "David - Personal" calendar, since the default calendar is my work calendar (Exchange-managed) Same sync settings as home PC.

   Test button shows format "g" with 0 entries between 11/4/2017 and 11/6/2017.  (which makes sense - no Outlook entries in the "David - Personal" calendar yet for today)

5) start Sync.  ERROR: GoogleCal entry ("Test entry #1") is not imported:
```
Sync started at 11/4/2017 11:49:00 PM
Syncing from 11/4/2017 to 11/6/2017
Outlook ↔ Google
--------------------------------------------------
Reading Outlook Calendar Entries...
0 Outlook calendar entries found.
--------------------------------------------------
Reading Google Calendar Entries...
1 Google calendar entries found.
--------------------------------------------------
Total inc. recurring items spanning sync date range...
Outlook 0, Google 1
--------------------------------------------------
0 Google calendar entries to be deleted.
0 Google calendar entries to be created.
--------------------------------------------------
0 Outlook calendar entries to be deleted.
0 Outlook calendar entries to be created.
--------------------------------------------------
Sync finished with success!
```
6) Let's try the other direction.  Create a new Outlook entry on the work PC (11:55pm - 11:58PM, "Test entry #2")

7) Start Sync.  PARTIAL SUCCESS AND FAILURE.  "Test Entry #2" is created on GoogleCal.  However OCGS on the work computer wants to delete "Test Entry #1" from GoogleCal.
```
Sync started at 11/4/2017 11:51:32 PM
Syncing from 11/4/2017 to 11/6/2017
Outlook ↔ Google
--------------------------------------------------
Reading Outlook Calendar Entries...
1 Outlook calendar entries found.
--------------------------------------------------
Reading Google Calendar Entries...
1 Google calendar entries found.
--------------------------------------------------
Total inc. recurring items spanning sync date range...
Outlook 1, Google 1
--------------------------------------------------
1 Google calendar entries to be deleted.
1 Google calendar entries to be created.
--------------------------------------------------
Deleting 1 Google calendar entries...
Not deleted: 11/4/2017 11:45 PM ="Test entry #1"
Done.
--------------------------------------------------
Creating 1 Google calendar entries...
Done.
--------------------------------------------------
0 Outlook calendar entries to be deleted.
0 Outlook calendar entries to be created.
--------------------------------------------------
Sync finished with success!
```
8) Back on the home PC, hit Sync.  FAILURE: "Test Entry #2" is not pulled in from GoogleCal:
```
Outlook ↔ Google
--------------------------------------------------
Reading Outlook Calendar Entries...
1 Outlook calendar entries found.
--------------------------------------------------
Reading Google Calendar Entries...
2 Google calendar entries found.
--------------------------------------------------
Total inc. recurring items spanning sync date range...
Outlook 1, Google 2
--------------------------------------------------
0 Google calendar entries to be deleted.
0 Google calendar entries to be created.
--------------------------------------------------
Comparing 1 existing Google calendar entries...
0 entries updated.
--------------------------------------------------
0 Outlook calendar entries to be deleted.
0 Outlook calendar entries to be created.
--------------------------------------------------
Comparing 1 existing Outlook calendar entries...
0 entries updated.
--------------------------------------------------
Sync finished with success!
```

9) Immediately hit Sync again on home PC.  This time, as with #7, home PC wants to delete the entry "Test Entry #1" from GoogleCal.  

10) Final test: create an entry in GoogleCal (via the web interface)  ("12/5/2017 12:30am - 12:45am, "Test Entry #3")

    Sync on work PC: SUCCESS - entry added to Outlook
    Sync on home PC: FAILURE - entry not added to Outlook - instead OGCS wants to delete "Test Entry #3" from GoogleCal.

So it appears that OGCS running on two different PCs, both of which use GoogleCal, is not working.  Is there any way to make this setup work for me?

Log files from the work PC are attached (home PC logs are similar)

[OGcalsync.log.2017-11-04.txt](https://github.com/phw198/OutlookGoogleCalendarSync/files/1443767/OGcalsync.log.2017-11-04.txt)
[OGcalsync.log.2017-11-05.txt](https://github.com/phw198/OutlookGoogleCalendarSync/files/1443768/OGcalsync.log.2017-11-05.txt)

**Comment by phw198** [comment]
Hi @dswhite42 - no, this looks like the configuration [3rd from the bottom](https://github.com/phw198/OutlookGoogleCalendarSync/raw/master/docs/images/supported_sync_configs.png); a common Google calendar being 2 way synced to from 2 different places. As you've guessed, this isn't supported and currently there's no workaround for that. Sorry!

**Comment by dswhite42** [comment]
Hmmm.  If deleting wasn't an issue (i.e. leave that feature unsupported), wouldn't it be possible to support that mode if all syncs were additive-only?  i.e. if calendar X has more events than Calendar Y, copy from X->Y .  If the opposite is true, copy from Y->X .  If they have different events, copy both directions until both calendars have the same events.   Then if Calendar Z syncs with calendar Y, do the same thing.

**Comment by phw198** [comment]
Copy what exactly? Each calendar item has to be matched based on a unique identifier - and that's the problem with having a common central Google calendar between two other calendars; is the identifier for Outlook calendar A (eg home) or Outlook calendar B (eg work) and so where should it add.delete/update. And how do options like Merge and Disable Delete impact etc etc.

I've toyed with the idea of creating a composite unique key that includes the computer name, but it's still quite a big change to the current sync engine...and the complexities this introduces are rather mind boggling. A challenge I'd like to crack at some point though...

**Comment by dswhite42** [comment]
I'm not familiar with the API for accessing eiher Outlook or Google, so
I can't make any specific code suggestions.  But in pseudocode I was   
thinking of something like this:

1. Each item in Outlook Calendar (home PC) is indexed by OUID, a concatenation of
start_date/end_date/title

1. Each item in Google  Calendar is indexed by GUID, a concatenation of
start_date/end_date/title

1. Pull all items from Outlook (home PC) and Google

1. If there are any OUIDS that do not have an equivalent GUID, push
those items from Outlook (home PC) to Google

1. If there are any GUIDS that do not have an equivalent GUID, pull
those items from Google to Outlook (home PC)

1. Repeat the same process using Outlook (work PC)

It seems that in this way, OGCS doesn't have to worry about how many   
clients are sync'ing to Google at all.  It's always just doing union   
operations.  

I know this is too simplistic and does not handle deletions at all.
But at least in my case, deleting is pretty rare, and if the worst   
thing that happens is that I have to delete the same appointment in 3
calendars, I can live with that.

**Comment by phw198** [comment]
:package: Development is now complete and [build v2.7.5.7.zip](https://github.com/phw198/OutlookGoogleCalendarSync/files/2300260/v2.7.5.7.zip) is now available for willing early testers before it goes to Alpha. Please [see instructions for applying](https://github.com/phw198/OutlookGoogleCalendarSync/wiki/Reporting-Problems#applying-a-hotfix) and make sure it is applied over at least v2.7.5.x.

Please let me know how it goes! If there are any issue, provide clear steps on how you have configured OGCS with your various calendars and the exact steps to replicate the problem.

**Comment by phw198** [comment]
This has been released to v2.7.6.
